File: src/heimdall/evaluator.py
from __future__ import annotations

import tempfile
from dataclasses import dataclass
from pathlib import Path

from .contract import EvalSpec, load_spec
from .reducer import reduce_evidence
from .report import render_markdown
from .runner import run


@dataclass(frozen=True)
class EvaluationArtifacts:
    evidence: dict[str, object]
    report: dict[str, object]
    markdown: str
    target_root: Path


def evaluate(manifest: str | Path) -> EvaluationArtifacts:
    spec: EvalSpec = load_spec(manifest)
    with tempfile.TemporaryDirectory(prefix="heimdall-") as temporary:
        evidence = run(spec, Path(temporary))
    report = reduce_evidence(evidence)
    return EvaluationArtifacts(
        evidence=evidence,
        report=report,
        markdown=render_markdown(report),
        target_root=spec.target_root,
    )

File: internal/evaluator/evaluator.go
package evaluator

import (
	"os"

	"github.com/JeremyDev87/heimdall/internal/contract"
	"github.com/JeremyDev87/heimdall/internal/reducer"
	"github.com/JeremyDev87/heimdall/internal/report"
	"github.com/JeremyDev87/heimdall/internal/runner"
)

type Artifacts struct {
	Evidence   map[string]any
	Report     map[string]any
	Markdown   string
	TargetRoot string
}

func Evaluate(manifest string) (*Artifacts, error) {
	spec, err := contract.LoadSpec(manifest)
	if err != nil {
		return nil, err
	}
	temporary, err := os.MkdirTemp("", "heimdall-")
	if err != nil {
		return nil, err
	}
	defer os.RemoveAll(temporary)
	evidence, err := runner.Run(spec, temporary)
	if err != nil {
		return nil, err
	}
	assessment, err := reducer.Reduce(evidence)
	if err != nil {
		return nil, err
	}
	return &Artifacts{Evidence: evidence, Report: assessment, Markdown: report.RenderMarkdown(assessment), TargetRoot: spec.TargetRoot}, nil
}
func ExitCode(state string) int {
	switch state {
	case "PASS":
		return 0
	case "FAIL":
		return 1
	case "BLOCKED":
		return 2
	case "INCONCLUSIVE":
		return 3
	default:
		return 2
	}
}
